xbmc/video/ContextMenus.cpp

```diff
@@ -168,4 +171,26 @@ bool CPlay::Execute(const CFileItemPtr& itemIn) const
   return true;
 };
 
+
+std::string CFavorite::GetLabel(const CFileItem& item) const
+{
+  //The 0 we're passing for contextWindow should only matter to folders
+  return g_localizeStrings.Get(XFILE::CFavouritesDirectory::IsFavourite(std::make_shared<CFileItem>(item).get(), 0) ? 14077 : 14076); // Add/Remove Favourite
+}
+
+bool CFavorite::IsVisible(const CFileItem& item) const
+{
+  //! @todo FAVOURITES Conditions on masterlock and localisation
+  return !item.IsParentFolder() && !item.IsPath("add") && !item.IsPath("newplaylist://") &&
+    !URIUtils::IsProtocol(item.GetPath(), "newsmartplaylist") && !URIUtils::IsProtocol(item.GetPath(), "newtag") &&
+    !URIUtils::IsProtocol(item.GetPath(), "musicsearch") &&
+    !StringUtils::StartsWith(item.GetPath(), "pvr://guide/") && !StringUtils::StartsWith(item.GetPath(), "pvr://timers/");
+}
+
+bool CFavorite::Execute(const CFileItemPtr& item) const
+{
+  return XFILE::CFavouritesDirectory::AddOrRemove(item.get(), 0);
```

Comment:
Btw, does this work? Usually a window is passed as seconds argument (for some reason). If not, the whole contextWindow thing can just be removed.